### DISCUSSION TITLE
restricting +rule button

### DISCUSSION THREAD

**Discussion by kuljeet2604** [react-querybuilder/react-querybuilder#633] [Q&A]
I have a use case where i need the  user to first click +group button and then ask him to add rule or nested group inside (to make ui more intuitive)
we can either achieve it by hiding the rule button on top most level and make it available only inside a group (i.e once user clicked +group button)
do you have a better alternative for it.

And if we need to adapt the first approach, what should be the changes to achieve hiding +rule button on top

**Comment by jakeboone02** [comment]
You can accomplish this with CSS or a custom "+ Rule" button. [This tips page in the documentation](https://react-querybuilder.js.org/docs/tips/limit-groups) should help with the concepts, so read that first.

The CSS method would be something like this:

```css
.queryBuilder > .ruleGroup > .ruleGroup-header > .ruleGroup-addRule {
  display: none;
}
```

That will hide the "+ Rule" button at the top level only.

If you want to be a little more certain that a "+ Rule" button doesn't show up at the top level, use a custom component like this:

```tsx
const CustomAddRuleAction = (props: ActionWithRulesAndAddersProps) => {
  // Don't render anything if this is the root group
  if (props.level === 0) {
    return null;
  }

  // Otherwise render as normal
  return <ActionElement {...props} />;
};

// Implementation:
<QueryBuilder
  fields={fields}
  query={query}
  onQueryChange={setQuery}
  controlElements={{
    addRuleAction: CustomAddRuleAction,
  }}
/>
```

Also make sure that at least one of the following are true:

1. The [`addRuleToNewGroups` prop](https://react-querybuilder.js.org/docs/components/querybuilder#addruletonewgroups) is undefined or `false`.
2. A `query` or `defaultQuery` prop is provided with no rules at the top level.

Finally, if you _really_ want to lock it down, use the [`onAddRule` prop](https://react-querybuilder.js.org/docs/components/querybuilder#onaddrule) in addition to the methods above, like this:

```tsx
// Return false if the rule is being added to the root group
const onAddRule = (r: RuleType, parentPath: number[]) =>
  parentPath.length === 0 ? false : r;

// Implementation:
<QueryBuilder
  fields={fields}
  query={query}
  onQueryChange={setQuery}
  controlElements={{
    addRuleAction: CustomAddRuleAction,
  }}
  onAddRule={onAddRule} // <--
/>
```

**Comment by kuljeet2604** [comment]
the above example of using a "CustomAddRuleAction" worked fine, but my addRule button styles is not getting applied, i am using QueryBuilderMantine.
rest all the styles are intact only the button lost its style.

attached screenshot for reference.
![Screenshot 2024-01-17 at 2 38 13 PM](https://github.com/react-querybuilder/react-querybuilder/assets/94591998/3ef5fa22-e4ad-4e69-9eff-2393ce081949)
![Screenshot 2024-01-17 at 2 38 06 PM](https://github.com/react-querybuilder/react-querybuilder/assets/94591998/b54661bd-08e6-415d-8ff2-0263252cda3d)

**Reply by jakeboone02** [reply]
```diff
-return <ActionElement {...props} />;
+return <MantineActionElement {...props} />;
```
